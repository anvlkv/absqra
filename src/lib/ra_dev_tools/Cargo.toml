[package]
name = "ra_dev_tools"
version = "0.1.0"
authors = ["a.nvlkv <a.nvlkv@protonmail.ch>"]
edition = "2018"
build = "build.rs"

[build-dependencies]
build-deps = "0.1.4"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = "0.16.0"
regex = "1.3.9"
example_files_macro = {path="./src/lib/example_files_macro"}
files_iterator = {path="./src/lib/files_iterator"}
