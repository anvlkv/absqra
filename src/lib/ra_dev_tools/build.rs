extern crate build_deps;

fn main() {
    // Enumerate files in sub-folder "examples/*", being relevant for the test-generation (as example)
    // If function returns with error, exit with error message.
    build_deps::rerun_if_changed_paths( "../../../examples/*/*" ).unwrap();

    // Adding the parent directory "examples" to the watch-list will capture new-files being added
    build_deps::rerun_if_changed_paths( "../../../examples/*" ).unwrap();
}