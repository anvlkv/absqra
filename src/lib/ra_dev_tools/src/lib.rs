pub extern crate insta;
pub use files_iterator::for_each_ra_example_file;

#[allow(unused_attributes)]
#[macro_use] pub use example_files_macro::make_example_tests;

#[cfg(test)]
mod test {
    use super::make_example_tests;

    #[make_example_tests]
    #[test]
    fn it_works_with(content: String, file_name: String) {
        println!("didn't panic with content: \n[{}] of file_name: {}", content, file_name);
    }
}