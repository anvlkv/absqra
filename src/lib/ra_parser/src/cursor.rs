use ra_lexer::token::{Token};
use ra_lexer::cursor::Position;
use core::iter::Peekable;

pub(crate) struct Cursor<'token, I> 
    where I: Iterator<Item = Token<'token>> 
{
    pub position: Position,
    pub level: u16,
    tokens: Peekable<I>,
    consumed_len: usize
}

impl <'token, I> Cursor<'token, I> 
where I: Iterator<Item = Token<'token>> 
{
    pub fn new(tokens: I) -> Self
        
    {
        Self {
            position: Position(0, 0),
            level: 0,
            tokens: tokens.peekable(),
            consumed_len: 0
        }
    }

    pub fn first_ahead(&mut self) -> Option<&Token<'token>> {
        self.tokens.peek()
    }

    pub fn bump(&mut self) -> Option<Token<'token>> {
        let tok = self.tokens.next();
        match tok {
            Some(token) => {
                self.position = token.position.1;
                self.level = token.level;
                self.consumed_len += 1;
                Some(token)
            }
            None => None
        }
    }

    pub fn is_eof(& mut self) -> bool {
        self.first_ahead() == None
    }

    // pub fn do_while<F, D>(&mut self, mut test: F, mut cb: D)
    //     where F: FnMut (u16, Position, TokenKind) -> bool,
    //     D: FnMut (Token)
    // {
    //     loop {
    //         let next_token = self.first_ahead();
    //         match next_token {
    //             Some(t) => {
    //                 if test(t.level, t.position.0, t.kind) {
    //                     cb(self.bump().unwrap())
    //                 }
    //                 else {
    //                     break
    //                 }
    //             },
    //             None => break
    //         }
    //     }
    // }

    // pub fn read_within(&mut self, open: TokenKind, close: TokenKind) -> Vec<Token> {
    //     let mut read: Vec<Token> = Vec::new();

    //     loop {
    //         let next_token = self.first_ahead();

    //         match next_token {
    //             Some(tok) => {
    //                 match tok.kind {
    //                     k if k == open => {
    //                         let open_token = self.bump().unwrap();
    //                         let mut nested = self.read_within(open_token.kind, close.clone());
    //                         read.append(&mut nested);
    //                     },
    //                     k if k == close => {
    //                         self.bump();
    //                         break
    //                     },
    //                     _ => read.push(self.bump().unwrap())
    //                 }
    //             },
    //             None => break
    //         }

    //         // println!("{:?}", &read);
    //     }

    //     read
    // }
}