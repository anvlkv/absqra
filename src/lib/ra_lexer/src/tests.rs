use crate::cursor::{Cursor, Position};
use crate::{tokenize, RaToken, TokenKind};

mod cursor {
    use super::{Cursor, Position};
    #[test]
    fn it_should_create() {
        Cursor::new("abc", Position(1, 0), 1, 0);
    }

    #[test]
    fn it_should_give_first_char() {
        let cur = Cursor::new("abc", Position(1, 0), 1, 0);
        assert_eq!(cur.first_ahead(), 'a');
    }

    #[test]
    fn it_should_give_second_char() {
        let cur = Cursor::new("abc", Position(1, 0), 1, 0);
        assert_eq!(cur.second_ahead(), 'b');
    }

    #[test]
    fn it_should_return_next_char() {
        let mut cur = Cursor::new("a", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
    }

    #[test]
    fn it_should_return_none_at_end_of_input() {
        let mut cur = Cursor::new("a", Position(1, 0), 1, 0);
        cur.bump();
        assert_eq!(cur.bump(), None);
    }

    #[test]
    fn it_should_confirm_eof() {
        let mut cur = Cursor::new("a", Position(1, 0), 1, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.is_eof(), true);
    }

    #[test]
    fn it_should_return_amount_of_consumed_symbols() {
        let mut cur = Cursor::new("abc", Position(1, 0), 1, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.len_consumed(), 3);
    }

    #[test]
    fn it_should_not_panic_when_encountering_funny_characters() {
        let mut cur = Cursor::new("ðŸš¬", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'ðŸš¬');
    }

    #[test]
    fn it_should_track_position() {
        let mut cur = Cursor::new("abc\nSOME", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.position, Position(1, 1));
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.position, Position(1, 2));
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.position, Position(1, 3));
        assert_eq!(cur.bump().unwrap(), 'S');
        assert_eq!(cur.position, Position(2, 1));
        assert_eq!(cur.bump().unwrap(), 'O');
        assert_eq!(cur.position, Position(2, 2));
        assert_eq!(cur.bump().unwrap(), 'M');
        assert_eq!(cur.position, Position(2, 3));
    }

    #[test]
    fn it_should_track_indent_level() {
        let mut cur = Cursor::new("a\n\tb\n\t\tc\nd", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.level, 1);
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.level, 2);
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.level, 3);
        assert_eq!(cur.position, Position(3, 3));
        assert_eq!(cur.bump().unwrap(), 'd');
        assert_eq!(cur.level, 1);
    }
}

mod lib {
    use super::{tokenize, Position, RaToken, TokenKind};

    #[test]
    fn it_should_create_iterator_of_tokens() {
        let mut stream = tokenize("abc");
        stream.next();
    }

    #[test]
    fn it_should_parse_multiple_lines() {
        let mut stream = tokenize("!?&/\n\tabc");
        stream.next();
        stream.next();
        stream.next();
        stream.next();
        assert_eq!(
            stream.next().unwrap().unwrap(),
            RaToken {
                kind: TokenKind::Identifier("abc"),
                content: "abc",
                position: (Position(2, 1), Position(2, 4)),
                len: 3,
                level: 2,
            }
        )
    }

    #[test]
    fn it_should_parse_identifiers() {
        let mut stream = tokenize("abc");
        assert_eq!(
            stream.next().unwrap().unwrap(),
            RaToken {
                kind: TokenKind::Identifier("abc"),
                content: "abc",
                position: (Position(1, 0), Position(1, 3)),
                len: 3,
                level: 1
            }
        );
    }

    use ra_dev_tools::insta::assert_json_snapshot;
    use ra_dev_tools::make_example_tests;

    #[make_example_tests]
    #[test]
    fn it_should_match_snapshots(contents: String, file_name: String) {
        let tokens: Vec<RaToken> = tokenize(&contents).map(|r| r.unwrap()).collect();
        assert_json_snapshot!(file_name, tokens)
    }

    mod strings {
        use super::{tokenize, Position, RaToken, TokenKind};
        #[test]
        fn it_should_parse_string_literals_with_double_quotes() {
            let mut stream = tokenize("\"some\"");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::StringLiteral("some"),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    level: 1
                }
            )
        }

        #[test]
        fn it_should_parse_string_literals_with_single_quotes() {
            let mut stream = tokenize("\'some\'");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::StringLiteral("some"),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    level: 1
                }
            )
        }
    }
    mod comments {
        use super::{tokenize, Position, RaToken, TokenKind};
        #[test]
        fn it_should_parse_single_line_comments() {
            let mut stream = tokenize("//abc");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 1
                }
            );
        }
        #[test]
        fn it_should_parse_multi_line_comments() {
            let mut stream = tokenize("/*abc\nSOME*/");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "abc\nSOME",
                    position: (Position(1, 0), Position(2, 6)),
                    len: 12,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_parse_multi_line_comments_with_fringe() {
            let mut stream = tokenize("/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "\n* abc\n* SOME\n",
                    position: (Position(1, 0), Position(4, 2)),
                    len: 18,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_parse_comments_followed_by_one_another() {
            let mut stream = tokenize("//some\n\n/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "some",
                    position: (Position(1, 0), Position(1, 6)),
                    len: 6,
                    level: 1
                }
            );

            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "\n* abc\n* SOME\n",
                    position: (Position(3, 0), Position(6, 2)),
                    len: 18,
                    level: 1
                }
            );
        }
    }
    
    mod numbers {
        use super::{tokenize, Position, RaToken, TokenKind};
        use crate::errors::LexerError;
        #[test]
        fn it_should_parse_numbers() {
            let mut stream = tokenize("123");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Int(123),
                    content: "123",
                    position: (Position(1, 0), Position(1, 3)),
                    len: 3,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Int(v) => assert_eq!(v, 123),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_separator() {
            let mut stream = tokenize("123,321");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123.321),
                    content: "123,321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123.321),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_separator() {
            let mut stream = tokenize("123.321");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123.321),
                    content: "123.321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123.321),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_thousands_separator() {
            let mut stream = tokenize("123.321,456");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123321.456),
                    content: "123.321,456",
                    position: (Position(1, 0), Position(1, 11)),
                    len: 11,
                    level: 1
                }
            );

            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123321.456),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_multiple_thousands_separator() {
            let mut stream = tokenize("123.321.123,456");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123321123.456),
                    content: "123.321.123,456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 1
                }
            );

            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123321123.456),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_and_multiple_thousands_separators() {
            let mut stream = tokenize("123,321,123.456");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123321123.456),
                    content: "123,321,123.456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 1
                }
            );

            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123321123.456),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_error_when_encountering_multiple_decimal_separators() {
            let mut stream = tokenize("123.321.123,456,654");
            match stream.next().unwrap() {
                Ok(_) => panic!("didn't error"),
                Err(e) => assert!(match e {
                    LexerError::UnexpectedCharacter(ch, _) => ch == ',',
                    _ => false
                })
            }
        }

        #[test]
        fn it_should_parse_negative_integer() {
            let mut stream = tokenize("-123");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Int(-123),
                    content: "-123",
                    position: (Position(1, 0), Position(1, 4)),
                    len: 4,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Int(v) => assert_eq!(v, -123),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_negative_float() {
            let mut stream = tokenize("-123.312");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(-123.312),
                    content: "-123.312",
                    position: (Position(1, 0), Position(1, 8)),
                    len: 8,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, -123.312),
                _ => panic!("unexpected kind")
            };
        }
    }

    mod content_block {
        use super::{tokenize, Position, RaToken, TokenKind};
        use crate::errors::LexerError;

        #[test]
        fn it_should_parse_content_blocks() {
            let mut stream = tokenize("`abc`");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::ContentBlock,
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_keep_inner_indents_when_parsing_content() {
            let mut stream = tokenize("`\n\tabc\n\t\tabc\n`");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::ContentBlock,
                    content: "\n\tabc\n\t\tabc\n",
                    position: (Position(1, 0), Position(4, 1)),
                    len: 14,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_error_when_content_block_is_not_closed_properly() {
            let mut stream = tokenize("`\n\tabc`");
            match stream.next().unwrap() {
                Ok(_) => panic!("parsed invalid content"),
                Err(e) => assert!(match e {
                    LexerError::UnexpectedEndOfInput(_) => true,
                    _ => false
                })
            }
        }
    }

    mod single_char_tokens {
        use super::{tokenize, Position, RaToken, TokenKind};
        #[test]
        fn it_should_parse_tokens() {
            let mut stream = tokenize("!?&/#");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Exclamation,
                    position: (Position(1, 0), Position(1, 1)),
                    content: "!",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Question,
                    position: (Position(1, 1), Position(1, 2)),
                    content: "?",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Ampersand,
                    position: (Position(1, 2), Position(1, 3)),
                    content: "&",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Slash,
                    position: (Position(1, 3), Position(1, 4)),
                    content: "/",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::HashPound,
                    position: (Position(1, 4), Position(1, 5)),
                    content: "#",
                    level: 1,
                    len: 1
                }
            );
        }
    }


}
