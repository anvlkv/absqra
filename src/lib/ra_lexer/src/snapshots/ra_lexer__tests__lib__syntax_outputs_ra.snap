---
source: src/lib/ra_lexer/src/tests.rs
expression: tokens
---
[
  {
    "kind": {
      "Identifier": "single_variable"
    },
    "len": 15,
    "content": "single_variable",
    "position": [
      [
        1,
        0
      ],
      [
        1,
        15
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 25,
    "content": " just an empty variable",
    "position": [
      [
        1,
        16
      ],
      [
        1,
        41
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 1
    },
    "len": 1,
    "content": "1",
    "position": [
      [
        3,
        0
      ],
      [
        3,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "len": 1,
    "content": "+",
    "position": [
      [
        3,
        2
      ],
      [
        3,
        3
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 1
    },
    "len": 1,
    "content": "1",
    "position": [
      [
        3,
        4
      ],
      [
        3,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 37,
    "content": " expression that has an output of 2",
    "position": [
      [
        3,
        6
      ],
      [
        3,
        43
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "len": 10,
    "content": "assignment",
    "position": [
      [
        5,
        0
      ],
      [
        5,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 39,
    "content": " expression that has value 1 assigned",
    "position": [
      [
        5,
        11
      ],
      [
        5,
        50
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 1
    },
    "len": 1,
    "content": "1",
    "position": [
      [
        6,
        0
      ],
      [
        6,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "len": 10,
    "content": "assignment",
    "position": [
      [
        8,
        0
      ],
      [
        8,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "len": 1,
    "content": "+",
    "position": [
      [
        8,
        11
      ],
      [
        8,
        12
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 2
    },
    "len": 1,
    "content": "2",
    "position": [
      [
        8,
        13
      ],
      [
        8,
        14
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 37,
    "content": " expression that has an output of 3",
    "position": [
      [
        8,
        15
      ],
      [
        8,
        52
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "nested_variable"
    },
    "len": 15,
    "content": "nested_variable",
    "position": [
      [
        10,
        0
      ],
      [
        10,
        15
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 38,
    "content": " expression that has output assigned",
    "position": [
      [
        10,
        16
      ],
      [
        10,
        54
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "len": 10,
    "content": "assignment",
    "position": [
      [
        11,
        0
      ],
      [
        11,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "len": 1,
    "content": "+",
    "position": [
      [
        11,
        11
      ],
      [
        11,
        12
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 2
    },
    "len": 1,
    "content": "2",
    "position": [
      [
        11,
        13
      ],
      [
        11,
        14
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 37,
    "content": " expression that has an output of 3",
    "position": [
      [
        11,
        15
      ],
      [
        11,
        52
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "len": 10,
    "content": "assignment",
    "position": [
      [
        13,
        0
      ],
      [
        13,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "Pipe",
    "len": 1,
    "content": "|",
    "position": [
      [
        13,
        11
      ],
      [
        13,
        12
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "nested_variable"
    },
    "len": 15,
    "content": "nested_variable",
    "position": [
      [
        13,
        13
      ],
      [
        13,
        28
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 24,
    "content": " logical expression OR",
    "position": [
      [
        13,
        29
      ],
      [
        13,
        53
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "len": 10,
    "content": "assignment",
    "position": [
      [
        15,
        0
      ],
      [
        15,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "Greater",
    "len": 1,
    "content": ">",
    "position": [
      [
        15,
        11
      ],
      [
        15,
        12
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "nested_variable"
    },
    "len": 15,
    "content": "nested_variable",
    "position": [
      [
        15,
        13
      ],
      [
        15,
        28
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 35,
    "content": " logical expression Comparison GT",
    "position": [
      [
        15,
        29
      ],
      [
        15,
        64
      ]
    ],
    "level": 1
  },
  {
    "kind": "OpenParentheses",
    "len": 1,
    "content": "(",
    "position": [
      [
        17,
        0
      ],
      [
        17,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 3
    },
    "len": 1,
    "content": "3",
    "position": [
      [
        17,
        1
      ],
      [
        17,
        2
      ]
    ],
    "level": 1
  },
  {
    "kind": "Asterisk",
    "len": 1,
    "content": "*",
    "position": [
      [
        17,
        2
      ],
      [
        17,
        3
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 5
    },
    "len": 1,
    "content": "5",
    "position": [
      [
        17,
        3
      ],
      [
        17,
        4
      ]
    ],
    "level": 1
  },
  {
    "kind": "CloseParentheses",
    "len": 1,
    "content": ")",
    "position": [
      [
        17,
        4
      ],
      [
        17,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "len": 1,
    "content": "+",
    "position": [
      [
        17,
        6
      ],
      [
        17,
        7
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 22
    },
    "len": 2,
    "content": "22",
    "position": [
      [
        17,
        8
      ],
      [
        17,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 46,
    "content": " grouped expression that has an output of 37",
    "position": [
      [
        17,
        11
      ],
      [
        17,
        57
      ]
    ],
    "level": 1
  },
  {
    "kind": "OpenParentheses",
    "len": 1,
    "content": "(",
    "position": [
      [
        19,
        0
      ],
      [
        19,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "len": 10,
    "content": "assignment",
    "position": [
      [
        19,
        1
      ],
      [
        19,
        11
      ]
    ],
    "level": 1
  },
  {
    "kind": "Pipe",
    "len": 1,
    "content": "|",
    "position": [
      [
        19,
        12
      ],
      [
        19,
        13
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "nested_variable"
    },
    "len": 15,
    "content": "nested_variable",
    "position": [
      [
        19,
        14
      ],
      [
        19,
        29
      ]
    ],
    "level": 1
  },
  {
    "kind": "CloseParentheses",
    "len": 1,
    "content": ")",
    "position": [
      [
        19,
        29
      ],
      [
        19,
        30
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "len": 1,
    "content": "+",
    "position": [
      [
        19,
        31
      ],
      [
        19,
        32
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 3
    },
    "len": 1,
    "content": "3",
    "position": [
      [
        19,
        33
      ],
      [
        19,
        34
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 65,
    "content": " grouped expression with logical expression, has an output of 6",
    "position": [
      [
        19,
        35
      ],
      [
        19,
        100
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 1
    },
    "len": 1,
    "content": "1",
    "position": [
      [
        21,
        0
      ],
      [
        21,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "len": 1,
    "content": "+",
    "position": [
      [
        21,
        2
      ],
      [
        21,
        3
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 1
    },
    "len": 1,
    "content": "1",
    "position": [
      [
        21,
        4
      ],
      [
        21,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "len": 1,
    "content": "+",
    "position": [
      [
        21,
        6
      ],
      [
        21,
        7
      ]
    ],
    "level": 1
  },
  {
    "kind": "OpenParentheses",
    "len": 1,
    "content": "(",
    "position": [
      [
        21,
        8
      ],
      [
        21,
        9
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 2
    },
    "len": 1,
    "content": "2",
    "position": [
      [
        21,
        9
      ],
      [
        21,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "Asterisk",
    "len": 1,
    "content": "*",
    "position": [
      [
        21,
        10
      ],
      [
        21,
        11
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 2
    },
    "len": 1,
    "content": "2",
    "position": [
      [
        21,
        11
      ],
      [
        21,
        12
      ]
    ],
    "level": 1
  },
  {
    "kind": "CloseParentheses",
    "len": 1,
    "content": ")",
    "position": [
      [
        21,
        12
      ],
      [
        21,
        13
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "len": 1,
    "content": "+",
    "position": [
      [
        21,
        14
      ],
      [
        21,
        15
      ]
    ],
    "level": 1
  },
  {
    "kind": "OpenParentheses",
    "len": 1,
    "content": "(",
    "position": [
      [
        21,
        16
      ],
      [
        21,
        17
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 3
    },
    "len": 1,
    "content": "3",
    "position": [
      [
        21,
        17
      ],
      [
        21,
        18
      ]
    ],
    "level": 1
  },
  {
    "kind": "Asterisk",
    "len": 1,
    "content": "*",
    "position": [
      [
        21,
        18
      ],
      [
        21,
        19
      ]
    ],
    "level": 1
  },
  {
    "kind": "OpenParentheses",
    "len": 1,
    "content": "(",
    "position": [
      [
        21,
        19
      ],
      [
        21,
        20
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 2
    },
    "len": 1,
    "content": "2",
    "position": [
      [
        21,
        20
      ],
      [
        21,
        21
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "len": 1,
    "content": "+",
    "position": [
      [
        21,
        21
      ],
      [
        21,
        22
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 2
    },
    "len": 1,
    "content": "2",
    "position": [
      [
        21,
        22
      ],
      [
        21,
        23
      ]
    ],
    "level": 1
  },
  {
    "kind": "CloseParentheses",
    "len": 1,
    "content": ")",
    "position": [
      [
        21,
        23
      ],
      [
        21,
        24
      ]
    ],
    "level": 1
  },
  {
    "kind": "CloseParentheses",
    "len": 1,
    "content": ")",
    "position": [
      [
        21,
        24
      ],
      [
        21,
        25
      ]
    ],
    "level": 1
  },
  {
    "kind": "Minus",
    "len": 1,
    "content": "-",
    "position": [
      [
        21,
        26
      ],
      [
        21,
        27
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Int": 18
    },
    "len": 2,
    "content": "18",
    "position": [
      [
        21,
        28
      ],
      [
        21,
        30
      ]
    ],
    "level": 1
  },
  {
    "kind": "Comment",
    "len": 40,
    "content": " bit more grouping, has an output of 0",
    "position": [
      [
        21,
        31
      ],
      [
        21,
        71
      ]
    ],
    "level": 1
  }
]
